gpt4all-chat/src/chatmodel.h

```diff
@@ -208,53 +260,77 @@ class ChatModel : public QAbstractListModel
         emit dataChanged(createIndex(index, 0), createIndex(index, 0), {ConsolidatedSourcesRole});
     }
 
     Q_INVOKABLE void updateThumbsUpState(int index, bool b)
     {
-        if (index < 0 || index >= m_chatItems.size()) return;
-
-        ChatItem &item = m_chatItems[index];
-        if (item.thumbsUpState != b) {
-            item.thumbsUpState = b;
-            emit dataChanged(createIndex(index, 0), createIndex(index, 0), {ThumbsUpStateRole});
+        bool changed = false;
+        {
+            QMutexLocker locker(&m_mutex);
+            if (index < 0 || index >= m_chatItems.size()) return;
+
+            ChatItem &item = m_chatItems[index];
+            if (item.thumbsUpState != b) {
+                item.thumbsUpState = b;
+                changed = true;
+            }
         }
+        if (changed) emit dataChanged(createIndex(index, 0), createIndex(index, 0), {ThumbsUpStateRole});
     }
 
     Q_INVOKABLE void updateThumbsDownState(int index, bool b)
     {
-        if (index < 0 || index >= m_chatItems.size()) return;
-
-        ChatItem &item = m_chatItems[index];
-        if (item.thumbsDownState != b) {
-            item.thumbsDownState = b;
-            emit dataChanged(createIndex(index, 0), createIndex(index, 0), {ThumbsDownStateRole});
+        bool changed = false;
+        {
+            QMutexLocker locker(&m_mutex);
+            if (index < 0 || index >= m_chatItems.size()) return;
+
+            ChatItem &item = m_chatItems[index];
+            if (item.thumbsDownState != b) {
+                item.thumbsDownState = b;
+                changed = true;
+            }
         }
+        if (changed) emit dataChanged(createIndex(index, 0), createIndex(index, 0), {ThumbsDownStateRole});
     }
 
     Q_INVOKABLE void updateNewResponse(int index, const QString &newResponse)
     {
-        if (index < 0 || index >= m_chatItems.size()) return;
-
-        ChatItem &item = m_chatItems[index];
-        if (item.newResponse != newResponse) {
-            item.newResponse = newResponse;
-            emit dataChanged(createIndex(index, 0), createIndex(index, 0), {NewResponseRole});
+        bool changed = false;
+        {
+            QMutexLocker locker(&m_mutex);
+            if (index < 0 || index >= m_chatItems.size()) return;
+
+            ChatItem &item = m_chatItems[index];
+            if (item.newResponse != newResponse) {
+                item.newResponse = newResponse;
+                changed = true;
+            }
         }
+        if (changed) emit dataChanged(createIndex(index, 0), createIndex(index, 0), {NewResponseRole});
     }
 
-    int count() const { return m_chatItems.size(); }
+    int count() const { QMutexLocker locker(&m_mutex); return m_chatItems.size(); }
+
+    ChatModelIterator begin() { return ChatModelIterator(this, 0); }
+    ChatModelIterator end() { return ChatModelIterator(this, m_chatItems.size()); }
```

Comment:
This should probably use QList<ChatItem>::iterator and not an index, otherwise it will definitely break if the size changes after calling end():
```suggestion
    ChatModelIterator begin() { return {this, m_chatItems.begin()}; }
    ChatModelIterator end() { return {this, m_chatItems.end()}; }
```